lima: add prime import support

enunes opened the pull request:
This patchset adds prime support to the lima kernel driver.
With this, it is possible to import buffers allocated in the display driver and exported with prime, enabling lima to render to a buffer that is being scanned to an output (such as HDMI).

Most of the initial commits in this series were not in linux 4.13 and were backported because they allow the sun4i-drm display driver to be used in the Allwinner A20, which is my test target. They define the reference device tree nodes that I used to port to the A20, along with a clock infrastructure change for the A20 which I found to be necessary to reduce the amount of work to make the sun4i-drm bindings to work on the A20.

This was finally tested with this application: https://github.com/enunes/mesa-test-programs/blob/master/gbm-bo-test.c
It is also necessary to have the mesa-lima patches to enable prime support there.

enunes committed:
lima: add prime import support

This enables lima to import a shared buffer from a dma_buf handle using
the DRM prime infrastructure.
With this, it is possible to allocate a dumb buffer with a display
driver to render to a display.
We can mostly use helper DRM callbacks, but gem_prime_import_sg_table
must have a custom function because we must allocate a lima_bo structure
and initialize custom fields other than the drm_gem_object.

This was tested on the Allwinner A20 with the sun4i-drm driver, along
with lima.

Signed-off-by: Erico Nunes <nunes.erico@gmail.com>

yuq commented on drivers/gpu/drm/lima/lima.h:
```diff
@@ -152,4 +152,6 @@ int lima_gem_wait(struct drm_file *file, u32 handle, u32 op, u64 timeout_ns);
 
 unsigned long lima_timeout_to_jiffies(u64 timeout_ns);
 
+struct drm_gem_object *lima_gem_prime_import(struct drm_device *dev, struct dma_buf *dma_buf);
```
Please move this declaration to the above "gem" chunk

yuq commented on drivers/gpu/drm/lima/lima_gem.c:
```diff
@@ -426,3 +427,81 @@ int lima_gem_wait(struct drm_file *file, u32 handle, u32 op, u64 timeout_ns)
 	return ret;
 }
+
+static struct lima_bo *lima_gem_alloc_object_dmabuf(struct drm_device *dev,
```
Seems we can move this into drm_driver.gem_create_object which is just for the cma helpers like vc4 driver

yuq replied:
Another better method I think is to define a gem_prime_import_sg_table of lima to do the dma_buf import without create a gem cma obj. In this way we can also support none-continuous dma_buf import.

yuq commented on drivers/gpu/drm/lima/lima_gem.c:
```diff
@@ -426,3 +427,81 @@ int lima_gem_wait(struct drm_file *file, u32 handle, u32 op, u64 timeout_ns)
+	return ERR_PTR(err);
+}
+
+struct drm_gem_object *lima_gem_prime_import(struct drm_device *dev,
```
we can just use the drm_gem_prime_import then
bo->dma_addr = sg_dma_address(sgt->sgl);
in this function with the above gem_bo create change

yuq commented on drivers/gpu/drm/lima/lima_drv.c:
```diff
@@ -200,7 +202,20 @@ static const struct file_operations lima_drm_driver_fops = {
+
+	.prime_handle_to_fd = drm_gem_prime_handle_to_fd,
+	.prime_fd_to_handle = drm_gem_prime_fd_to_handle,
+	.gem_prime_export   = drm_gem_prime_export,
```
Have you tested the export function? I believe it'll just crash if user call it. If not implement correctly, you can just not assign this function pointer so that user won't call into it.

yuq replied:
Also if drop this function, gem_prime_get_sg_table is not needed too.

yuq replied:
So does the gem_prime_vmap/gem_prime_vunmap/gem_prime_mmap/prime_handle_to_fd

yuq reviewed (changes requested):
I know we can just use the cma helpers to unify the ops of mem allocated from lima and mem imported to lima, but the reason why I didn't use it is the mali400 GPU has MMU that can handle none-continuous mem. I plan to move to a more advanced MM system like TTM in the future.

yuq commented:
If you want to test the prime export function, just do opposite in your gbm-bo-test:
1. create gbm_bo of lima with GBM_BO_USE_RENDERING|GBM_BO_USE_SCANOUT|GBM_BO_USE_LINEAR
2. lima export the bo to fd with gbm_bo_get_fd
3. make display driver import the fd

enunes commented:
Thanks for the detailed and fast review. I used some code that I found in other drivers and this was basically a cleaned version of the first thing that worked.

If you are okay with the other backported patches for the A20, maybe I can submit that as a separate pull request and drop them from this one.

I don't know if it makes sense to do exporting from lima in the sun4i-drm case. I understand that the mali has a MMU but I don't know if the display driver would be able to use it if I exported a non-contiguous buffer allocated with lima.
I'm planning to drop the gem_prime_export from the next revision of this patch to make it simpler, possibly do it in a further patchset if it will be nice to have it.

yuq commented:
Yes, I'm OK with the other patches, you can just create a new pull request for others.

Prime export is used by xserver and wayland to pass bo as fd and use case like I described above, so we'll need it anyway in the future. But you can drop it for now.

The display driver will need the export dma_buf be continuous, so the gbm_bo create hint flag GBM_BO_USE_SCANOUT can be used to tell lima driver to create a continuous bo or we can realloc the mem when export bo. But this indeed create more complexity than all cma case.

enunes commented:
I have updated the branch, removed the A20 related patches from here and created pull request #3. I implemented what I understood from your suggestion and I think it is better now.
So most of the code was not really necessary, indeed the drm helper for gem_prime_import will do the right thing if we define a custom lima_gem_prime_import_sg_table to allocate and initialize a lima_bo.
I'm not sure if this is enough to support non-contiguous allocations yet, since it still uses just sg_dma_address.
Let me know if there is still something that can be improved.

yuq commented on drivers/gpu/drm/lima/lima.h:
```diff
@@ -149,6 +149,8 @@ int lima_gem_submit(struct drm_file *file, struct lima_sched_pipe *pipe,
 		    void *frame, u32 *fence);
 int lima_gem_wait(struct drm_file *file, u32 handle, u32 op, u64 timeout_ns);
+struct drm_gem_object *lima_gem_prime_import_sg_table(struct drm_device *dev,
+		    struct dma_buf_attachment *attach, struct sg_table *sgt);
```
Please align with the "(" or re-start a new line for all args.

yuq commented on drivers/gpu/drm/lima/lima_gem.c:
```diff
@@ -53,10 +55,15 @@ static struct lima_bo *lima_gem_create_bo(struct drm_device *dev, u32 size, u32
-	if (!bo->cpu_addr) {
-		err = -ENOMEM;
-		goto err_out1;
+	if (dma_addr)
```
What about the bo free case, we can't do dma_free for bo imported from other device.

enunes replied:
True, do you have a suggestion on how to handle this? Is there some readily available reference counting that we can use to free this, or do we need to store some additional information in lima_bo to handle this?

yuq replied:
From the vc4_free_object of vc4 driver, we can use the gem_bo->import_attach to distinguish a self created bo and a imported bo.

yuq replied:
There's another place which is not shown in this patch that the bo->resv should also be from the dma_buf obj. Please reference the vc4_prime_import_sg_table. This resv is used to sync dma_buf read/write between multi users. In the display/render case, render will queue a write fence to this resv obj, display will wait the resv obj (so the write fence) before display. So the lima_bo should use the same resv as dma_buf.

yuq commented on drivers/gpu/drm/lima/lima_gem.c:
```diff
@@ -426,3 +433,16 @@ int lima_gem_wait(struct drm_file *file, u32 handle, u32 op, u64 timeout_ns)
 }
+
+struct drm_gem_object *lima_gem_prime_import_sg_table(struct drm_device *dev,
+		    struct dma_buf_attachment *attach, struct sg_table *sgt)
```
Same alignment requirement

yuq reviewed (changes requested):
This looks much better. As for the none-continuous mem support, this patch won't do it, because the MM of lima need to support it first.

enunes commented:
Fixed the alignment in the modified functions and made the dma_free_coherent conditional on obj->import_attach.

yuq commented on drivers/gpu/drm/lima/lima.h:
```diff
@@ -149,6 +149,10 @@ int lima_gem_submit(struct drm_file *file, struct lima_sched_pipe *pipe,
 		    void *frame, u32 *fence);
 int lima_gem_wait(struct drm_file *file, u32 handle, u32 op, u64 timeout_ns);
+struct drm_gem_object *
+lima_gem_prime_import_sg_table(struct drm_device *dev,
```
Sorry for my miss leading comments, I really mean new line arg align with the start of the last line arg not the "(", so in this case, "struct" align with the "struct". Please refer to my other functions which have multi line of arg.

yuq commented on drivers/gpu/drm/lima/lima_gem.c:
```diff
@@ -34,7 +35,8 @@ struct lima_drm_priv *to_lima_drm_priv(struct drm_file *file)
 }
 
-static struct lima_bo *lima_gem_create_bo(struct drm_device *dev, u32 size, u32 flags)
+static struct lima_bo *lima_gem_create_bo(struct drm_device *dev, u32 size,
```
Seems besides dma_addr, we have to handle a new resv obj arg. So I think we can move the amd_addr and resv init out of this function to the corresponding lima_gem_create_handle & lima_gem_prime_import_sg_table.

yuq reviewed (changes requested):
We also need to handle bo->resv.

enunes commented:
I changed the bo->resv handling according to what is done in vc4. Apparently etnaviv also does the same. Moved dma_addr and resv pointer initialization to lima_gem_prime_import_sg_table and lima_gem_create_handle accordingly.

I'm still not sure about the alignment requirement, so I just made it like the other functions in this file. I was trying to follow what I saw in some drm header before. Some lines are now over 80 characters, I assume that is fine since there are other lines like this in this file.

yuq reviewed (commented):
This patch looks good for me. Just in case, have you done a test that read out the imported bo from lima side (like dump the imported bo with glReadPixels)? Because I see the imported bo won't have cpu_addr, not sure if it will be OK in lima_gem_mmap which will call dma_mmap_coherent with this cpu_addr (it should depend on the implementation of dma_mmap_coherent of different arch). I think we will switch to new MM in the future, so no need to address this problem if the test is OK at least on A20.

enunes commented:
I have tested gbm-surface from your gfx repo (reads buffer, but it is not an imported buffer) and gbm-bo-test from my mesa-test-programs repo (imported buffer but render only, no reads). I did not test a mix of this, as in reading from an imported buffer. I can fix it, but I don't know what exactly to fill to cpu_addr in this case.

enunes commented:
Actually I just re-tested and it seems that reading from the imported framebuffer works as it is now. gbm-bo-test was even doing a glReadPixels but just not writing it to any image file.

yuq commented:
OK, that's fine, as the read works on A20, I think we can just do it this way. After new MM, we should have no this problem.
